NAMESPACE

```diff
@@ -13,6 +13,7 @@ export(get_warnings)
 export(join)
 export(new_qenv)
 export(qenv)
+export(subset)
```

Comment:
We need to come up with a different name, since `subset` is an S3 function in `base`

```r
> library(teal.code)

Attaching package: ‘teal.code’

The following object is masked from ‘package:base’:

    subset
```

Reply:
propositions for new name

- trim
- limit
- filter
- reduce
- shrink
- crop
- select
- slice
- truncate
- prune

Reply:
If you want S3 method for qenv you just need to make S3 methods `subset.qenv <- `, `subset.qenv.error <-`. Since we are talking about different names, what is the most relevant name for this? What would you like to use to trim qenv object?

Reply:
I would keep all the methods as S4 so it's consistent across other methods that we have for `qenv`.
I think `subset`, `truncate`, `trim` of `filter` are the most relevant in here.

Reply:
Let's consider how we subset slots:
- we subset `@env` with `@env[<names>]`
- we subset `@join_keys` with `@env[<names>]`

`[.qenv` also makes sense IMO

Reply:
`@env` is not subsettable

```r
> x <- new.env(); x$a <- 1; x$b <- 2
> x[c('a', 'b')]
Error in x[c("a", "b")] : object of type 'environment' is not subsettable
> x[[c('a', 'b')]]
Error in x[[c("a", "b")]] : wrong arguments for subsetting an environment
```

`[.` works only for names of length 1, that's why we use `mget` in the implementation.

Because of this reason I would strive from `[.`, since environments are not subsettable

Reply:
@m7pr How would you address subsetting of `teal_data` object in the future? Specifically `join_keys` slot?

Reply:
@gogonzo `join_keys` is just an S3 list with custom `attributes` and a customized `print` method. So using `[.` or `[[.` is natural in this case.

`teal_data` and `qenv` are a bit more complex objects as they have multiple fields of different classes.
`teal_data` doesn't handle `[[.` for multiple names, just for one and returns the dataset. `teal_code` does the same with `[[.` being a synonym to `get_var`


```r
data <- teal_data(iris = iris, mtcars = mtcars)
data[['iris']] # works 
data[[c('iris', 'mtcars')]] # doesn't work
first argument has length > 1
```

Both `teal_data` and `qenv` already have defined methods for `[[.` that do `get_var`

```r
> getMethod("[[","qenv")
Method Definition:

function (x, i, j, ...) 
{
    .local <- function (x, i) 
    {
        get_var(x, i)
    }
    .local(x, i, ...)
}
<bytecode: 0x000001cff0ac0f78>
<environment: namespace:teal.code>

Signatures:
        x     
target  "qenv"
defined "qenv"
```


So the idea is to introduce a method with a different name than `[[.`, for example `truncate` that returns trimmed `qenv` and respected `@code`, `@env`, `@warnings` and `@messages` fields. For `teal_data` we can introduce the same method with the same name that also handles `@verified`, `@datanames` and `@join_keys` fields if needed.

Reply:
> join_keys is just an S3 list with custom attributes and a customized print method. So using [. or [[. is natural in this case.

What "natural" means? There is `[.join_keys` overriden in the `teal.data` so it is not entirely natural.

> teal_data and qenv are a bit more complex objects as they have multiple fields of different classes.
teal_data doesn't handle [[. for multiple names, just for one and returns the dataset. teal_code does the same with [[. being a synonym to get_var

`[[` doesn't handle multiple index in general. For example `list("a", "b")[[c(1, 2)]]` fails. In general `[` and `[[` are completely different functions - `[` subsets and keeps the original class while `[[` extracts element.  I suggested possible usage of `[` for `qenv` which in certain design wouldn't require introducing extra `[.teal_data`.

Reply:
Alrighty @gogonzo , so `[[.` exists for `qenv` and extracts one element. `[.` does now exist yest for `qenv` but could exist and could extract multiple elements. I think it make sense. Thanks for the clarification. I will continue with the proposition to use `[.`